Framework/Algorithms/src/SANSSolidAngle.cpp

```diff
@@ -0,0 +1,184 @@
+      return cosAlpha * cosAlpha * cosAlpha;
+    };
+  } else {
+    throw std::runtime_error("Invalid type of correction");
+  }
+}
+
+/**
+ * Compute the solid angle using the correct angular function
+ */
+class CalculateSolidAngle {
+public:
+  CalculateSolidAngle(const double pixelArea, const SpectrumInfo &spectrumInfo,
+                      const ComponentInfo &componentInfo,
+                      std::function<double(int64_t)> &&f)
+      : m_pixelArea(pixelArea), m_spectrumInfo(spectrumInfo),
+        m_componentInfo(componentInfo), m_f(std::move(f)) {}
+  double operator()(int64_t histogramIndex) const {
+    Kernel::V3D scaleFactor = m_componentInfo.scaleFactor(histogramIndex);
+    double scaledPixelArea = m_pixelArea * scaleFactor[0] * scaleFactor[1];
+    double angularTerm = m_f(histogramIndex);
+    double distance = m_spectrumInfo.l2(histogramIndex) * 1.e3; // mm
+    return (scaledPixelArea * angularTerm) / (distance * distance);
+  }
+
+private:
+  const double m_pixelArea;
+  const SpectrumInfo &m_spectrumInfo;
+  const ComponentInfo &m_componentInfo;
+  const std::function<double(int64_t)> m_f;
+};
+
+void SANSSolidAngle::exec() {
+
+  // Get the workspaces
+  API::MatrixWorkspace_const_sptr inputWS = getProperty("InputWorkspace");
+  const int64_t numberOfSpectra =
+      static_cast<int64_t>(inputWS->getNumberHistograms());
+  API::MatrixWorkspace_sptr outputWS =
+      WorkspaceFactory::Instance().create(inputWS, numberOfSpectra, 2, 1);
+  // The result of this will be a distribution
+  outputWS->setDistribution(true);
+  outputWS->setYUnit("");
+  outputWS->setYUnitLabel("Steradian");
+  setProperty("OutputWorkspace", outputWS);
+
+  const auto &spectrumInfo = inputWS->spectrumInfo();
+  const auto &componentInfo = inputWS->componentInfo();
+  const double pixelSizeX =
+      inputWS->getInstrument()->getNumberParameter("x-pixel-size")[0];
+  const double pixelSizeY =
+      inputWS->getInstrument()->getNumberParameter("y-pixel-size")[0];
+
+  g_log.debug() << "Pixel sizes (mm) X=" << pixelSizeX << " Y=" << pixelSizeY
+                << '\n';
+
+  const std::string type = getProperty("Type");
+  auto angularFunction = getAngularFunction(spectrumInfo, type);
+  const CalculateSolidAngle calculateSolidAngle(pixelSizeX * pixelSizeY,
+                                                spectrumInfo, componentInfo,
+                                                std::move(angularFunction));
+
+  Progress prog(this, 0.0, 1.0, numberOfSpectra);
+
+  // Loop over the histograms (detector spectra)
+  PARALLEL_FOR_IF(Kernel::threadSafe(*outputWS, *inputWS))
+  for (int64_t i = 0; i < numberOfSpectra; ++i) {
+    PARALLEL_START_INTERUPT_REGION
+    if (spectrumInfo.hasDetectors(i) && !spectrumInfo.isMonitor(i)) {
+      // Copy over the spectrum number & detector IDs
+      outputWS->getSpectrum(i).copyInfoFrom(inputWS->getSpectrum(i));
+      outputWS->mutableX(i) = inputWS->x(i);
```

Comment:
In fact, this is also wrong. The input might as well be a TOF workspace with many bins, but the output should have only one number; that is, the solid angle value.